### PR TITLE
enhancement(monitor): enable dropping events at the kernel level

### LINKED ISSUE #1088: Ability to enable/disable visibility at runtime using annotations.

### LINKED ISSUE #1089: Ability to enable/disable visibility at host, namespace
There should be a way to set default cluster-wide visibility.

maps to use to acheive host/namespace visibility controls

Image

Ability to enable/disable visibility at deployment level has been moved to a seperate issue

### PR DESCRIPTION
**Pull request by achrefbensaad**
Signed-off-by: Achref ben saad <achref@accuknox.com>

**Purpose of PR?**: #1088 #1089 #1006

**Does this PR introduce a breaking change?** No

**If the changes in this PR are manually verified, list down the scenarios covered:**:

**Additional information for reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_

**Checklist:**
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [x] Commit has integration tests

### CODE REVIEW

File: KubeArmor/monitor/systemMonitor.go [hunk 1]
```diff
@@ -31,6 +33,9 @@ import (
 const (
 	PermissionDenied = -13
 	MaxStringLen     = 4096
+	PinPath          = "/sys/fs/bpf"
+	visibility_off   = uint32(1)
+	visibility_on    = uint32(0)
```
Comment (daemon1024) [inline KubeArmor/monitor/systemMonitor.go]: let's handle linter errors

File: KubeArmor/BPF/system_monitor.c [hunk 2]
```diff
@@ -429,50 +468,62 @@ static __always_inline u32 remove_pid_ns()
...
+
+    if (!on_off_switch) {
+        return trace_syscall;
     }
 
+    if (*on_off_switch){
+        return ignore_syscall;
+    }
+#if defined(MONITOR_HOST)
+
```
Comment (daemon1024) [inline KubeArmor/BPF/system_monitor.c:534]: Do we need this anymore? I think we use 0,0 to signify that right? or will it be handled in a separate PR?

File: KubeArmor/BPF/system_monitor.c [hunk 3]
```diff
@@ -273,6 +273,35 @@ BPF_PERCPU_ARRAY(bufs_offset, u32, 3);
...
+// == Visibility == //
+
+const u32 file_probe = 0;
+const u32 process_probe = 1;
+const u32 network_probe = 2;
+const u32 caps_probe = 3;
+
+const u32 trace_syscall = 0;
+const u32 ignore_syscall = 1;
+
```
Comment (daemon1024) [inline KubeArmor/BPF/system_monitor.c]: Should we perhaps use macro constants/ enums here, since that is the pattern with other constants? WDYT?

File: KubeArmor/BPF/system_monitor.c [hunk 4]
```diff
@@ -346,6 +375,16 @@ static struct file *get_task_file(struct task_struct *task)
...
 
+static __always_inline void get_outer_key(struct outer_key *pokey,
+                                 struct task_struct *t, u32 *pid_ns) {
+  pokey->pid_ns = *pid_ns;
+  pokey->mnt_ns = get_task_mnt_ns_id(t);
+  if (pokey->pid_ns == PROC_PID_INIT_INO) {
+    pokey->pid_ns = 0;
+    pokey->mnt_ns = 0;
+  }
+}
```
Comment (daemon1024) [inline KubeArmor/BPF/system_monitor.c]: why are we extracting pid_ns from the task structure outside of the function?

**Review by daemon1024 — CHANGES_REQUESTED** [review]
LGTM in general. Some minor nits inline and queries inline.

File: KubeArmor/monitor/systemMonitor.go [hunk 5]
```diff
@@ -209,6 +239,150 @@ func isIgnored(item string, ignoreList []string) bool {
...
+		mon.Logger.Printf("Successfully deleted visibiliy map with key=%+v from the kernel", nsKey)
+	}
+}
+
+// UpdateHostVisibility Function
+func (mon *SystemMonitor) UpdateHostVisibility() {
+	nsKey := NsKey{
+		PidNS: 0,
+		MntNS: 0,
+	}
```
Comment (daemon1024) [inline KubeArmor/monitor/systemMonitor.go:373]: Since we already have this code. We should eliminate the need for conditional compilation. and update the in kernel logic to not consider it as well.

**Review by daemon1024 — CHANGES_REQUESTED** [review]

**Comment by achrefbensaad** [comment]
New logic for droping events:
image

File: KubeArmor/BPF/system_monitor.c [hunk 6]
```diff
@@ -1435,7 +1504,7 @@ int kprobe__chown(struct pt_regs *ctx)
 SEC("kretprobe/__x64_sys_chown")
 int kretprobe__chown(struct pt_regs *ctx)
 {
-    return trace_ret_generic(_SYS_CHOWN, ctx, ARG_TYPE0(FILE_TYPE_T)|ARG_TYPE1(INT_T)|ARG_TYPE2(INT_T));
+    return trace_ret_generic(_SYS_CHOWN, ctx, ARG_TYPE0(FILE_TYPE_T)|ARG_TYPE1(INT_T)|ARG_TYPE2(INT_T), _FILE_PROBE);
```
Comment (nyrahul) [inline KubeArmor/BPF/system_monitor.c]: Why is chown a FILE_PROBE but setuid a CAP_PROBE?

File: KubeArmor/core/kubeUpdate.go [hunk 7]
```diff
@@ -2569,6 +2571,75 @@ func validateGlobalDefaultPosture(posture string) string {
...
+	} else {
+		nsPatch := corev1.Namespace{}
+		nsPatch.Annotations = make(map[string]string)
+		nsPatch.Annotations[key] = DefaultVisibility
+		if kl.ContainsElement(dm.SystemMonitor.UntrackedNamespaces, ns.Name) {
+			nsPatch.Annotations[key] = "none"
+		}
+		patch, err := json.Marshal(nsPatch)
+		if err != nil {
+			dm.Logger.Warnf("Cannot marshal namespace patch for %s, err=%s", ns.Name, err)
```
Comment (nyrahul) [inline KubeArmor/core/kubeUpdate.go:2599]: return from here, since otherwise it will result in a crash.

File: KubeArmor/core/kubeUpdate.go [hunk 8]
```diff
@@ -2569,6 +2571,75 @@ func validateGlobalDefaultPosture(posture string) string {
...
+		}
+		_, err = K8s.K8sClient.CoreV1().Namespaces().Patch(context.Background(), ns.Name, types.MergePatchType, patch, metav1.PatchOptions{})
+		if err != nil {
+			dm.Logger.Warnf("Cannot patch namespace %s, err=%s", ns.Name, err)
+		}
+	}
+	return false
+}
```
Comment (nyrahul) [inline KubeArmor/core/kubeUpdate.go]:
```suggestion
func (dm *KubeArmorDaemon) validateVisibility(scope string, ns *corev1.Namespace, DefaultVisibility string) bool {
	key := "kubearmor-visibility"
	if ns.Annotations != nil && ns.Annotations[key] != "" {
		if strings.Contains(ns.Annotations[key], scope) {
			return true
		}
		return false
	}
	nsPatch := corev1.Namespace{}
	nsPatch.Annotations = make(map[string]string)
	nsPatch.Annotations[key] = DefaultVisibility
	if kl.ContainsElement(dm.SystemMonitor.UntrackedNamespaces, ns.Name) {
		nsPatch.Annotations[key] = "none"
	}
	patch, err := json.Marshal(nsPatch)
	if err != nil {
		dm.Logger.Warnf("Cannot marshal namespace patch for %s, err=%s", ns.Name, err)
		return false
	}
	_, err = K8s.K8sClient.CoreV1().Namespaces().Patch(context.Background(), ns.Name, types.MergePatchType, patch, metav1.PatchOptions{})
	if err != nil {
		dm.Logger.Warnf("Cannot patch namespace %s, err=%s", ns.Name, err)
	}
	return false
}
```

**Review by nyrahul — CHANGES_REQUESTED** [review]

File: KubeArmor/monitor/systemMonitor.go [hunk 9]
```diff
@@ -215,6 +245,150 @@ func isIgnored(item string, ignoreList []string) bool {
...
+			Name:       "kubearmor_visibility",
+			Type:       cle.HashOfMaps,
+			KeySize:    8,
+			ValueSize:  4,
+			MaxEntries: 65535,
+			Pinning:    cle.PinByName,
+			InnerMap:   &mon.BpfVisibilityMapSpec,
+		}, cle.MapOptions{
+			PinPath: PinPath,
+		})
```
Comment (daemon1024) [inline KubeArmor/monitor/systemMonitor.go]: PinPath needs to be handled via how we mount bpffs. Can you refactor the implementation in #1097 to accomodate this usecase as well?

**Review by daemon1024 — CHANGES_REQUESTED** [review]

File: KubeArmor/core/kubeUpdate.go (hunk 8 above)
Comment (nyrahul) [reply KubeArmor/core/kubeUpdate.go]: This was code factoring point. We can ignore this point.

File: KubeArmor/BPF/system_monitor.c (hunk 6 above)
Comment (achrefbensaad) [reply KubeArmor/BPF/system_monitor.c]: Assignements were discussed via internal slack messages.

**Review by nyrahul — CHANGES_REQUESTED** [review]
Please fix the lint issues. thanks

**Review by daemon1024 — APPROVED** [review]
Verified Changes. LGTM

File: KubeArmor/monitor/systemMonitor.go [hunk 10]
```diff
@@ -215,6 +249,150 @@ func isIgnored(item string, ignoreList []string) bool {
...
+
+	return err
+}
+
+// DestroyBPFMaps Function
+func (mon *SystemMonitor) DestroyBPFMaps() {
+	err := mon.BpfNsVisibilityMap.Unpin()
+	if err != nil {
+		mon.Logger.Warnf("error unpinning bpf map kubearmor_visibility %v", err)
+	}
```
Comment (Ankurk99) [inline KubeArmor/monitor/systemMonitor.go:277]: return

File: KubeArmor/monitor/systemMonitor.go [hunk 11]
```diff
@@ -215,6 +249,150 @@ func isIgnored(item string, ignoreList []string) bool {
...
+	err = mon.BpfNsVisibilityMap.Close()
+	if err != nil {
+		mon.Logger.Warnf("error closing bpf map kubearmor_visibility %v", err)
+	}
```
Comment (Ankurk99) [inline KubeArmor/monitor/systemMonitor.go:281]: return

File: KubeArmor/common/bpffs.go [hunk 12]
```diff
@@ -52,23 +54,23 @@ var (
...
 // We also check and error if there have been multiple mounts at
 // the same point. See - https://patchwork.kernel.org/project/netdevbpf/patch/20220223131833.51991-1-laoar.shao@gmail.com/
-func (be *BPFEnforcer) CheckOrMountBPFFs(bpfRoot string) {
+func CheckOrMountBPFFs(bpfRoot string) {
 	mountOnce.Do(func() {
-		if err := be.checkOrMountBPFFs(bpfRoot); err != nil {
-			be.Logger.Err("Unable to mount BPF filesystem")
+		if err := checkOrMountBPFFs(bpfRoot); err != nil {
+			kg.Err("Unable to mount BPF filesystem")
 		}
```
Comment (Ankurk99) [inline KubeArmor/common/bpffs.go:61]: return

File: KubeArmor/core/kubeUpdate.go [hunk 13]
```diff
@@ -2607,8 +2684,14 @@ func (dm *KubeArmorDaemon) WatchDefaultPosture() {
 					CapabilitiesAction: cp,
 				}
 				annotated := fa || na || ca
+				visibility := tp.Visibility{
+					File:         dm.validateVisibility("file", ns, cfg.GlobalCfg.Visibility),
+					Process:      dm.validateVisibility("process", ns, cfg.GlobalCfg.Visibility),
+					Network:      dm.validateVisibility("network", ns, cfg.GlobalCfg.Visibility),
+					Capabilities: dm.validateVisibility("capabilities", ns, cfg.GlobalCfg.Visibility),
```
Comment (Ankurk99) [inline KubeArmor/core/kubeUpdate.go:2691]: can we directly define the visibility flag in the validatevisibility() func directly, so that we are not calling it every time?

File: KubeArmor/enforcer/bpflsm/enforcer.go [hunk 14]
```diff
@@ -136,7 +134,7 @@ func NewBPFEnforcer(node tp.Node, logger *fd.Feeder) (*BPFEnforcer, error) {
 
 	if err := loadEnforcer_pathObjects(&be.objPath, &ebpf.CollectionOptions{
 		Maps: ebpf.MapOptions{
-			PinPath: GetMapRoot(),
+			PinPath: common.GetMapRoot(),
 		},
 	}); err != nil {
 		be.Logger.Warnf("error loading BPF LSM Path objects. This usually suggests that the system doesn't have the system has `CONFIG_SECURITY_PATH=y`: %v", err)
```
Comment (Ankurk99) [inline KubeArmor/enforcer/bpflsm/enforcer.go:140]: return

**Review by Ankurk99 — CHANGES_REQUESTED** [review]
Have some suggestions, ptal

**Comment by daemon1024** [comment]
Noting the init container changes are not backwards compatible, so we  need to establish stable and version init containers as well.

File: KubeArmor/core/kubeUpdate.go (hunk 13 above)
Comment (daemon1024) [reply KubeArmor/core/kubeUpdate.go:2691]: I am not sure what do you mean by this, Can you explain?

File: KubeArmor/common/bpffs.go (hunk 12 above)
Comment (daemon1024) [reply KubeArmor/common/bpffs.go:61]: No executions after that do I don't think it's needed

File: KubeArmor/enforcer/bpflsm/enforcer.go (hunk 14 above)
Comment (daemon1024) [reply KubeArmor/enforcer/bpflsm/enforcer.go:140]: We don't want to, path objects are optional. It's fine if they fail to load and other enforcers are functioning as expected

File: KubeArmor/monitor/systemMonitor.go (hunk 10 above)
Comment (daemon1024) [reply KubeArmor/monitor/systemMonitor.go:277]: We want to continue to try to close it even if unpin fails

File: KubeArmor/monitor/systemMonitor.go (hunk 11 above)
Comment (daemon1024) [reply KubeArmor/monitor/systemMonitor.go:281]: No executions after it so doesn't matter imo?

**Review by daemon1024 — COMMENTED** [review]
@Ankurk99 addressed comments inline

File: KubeArmor/core/kubeUpdate.go (hunk 13 above)
Comment (Ankurk99) [reply KubeArmor/core/kubeUpdate.go:2691]: File: dm.validateVisibility("file", ns) and keep cfg.GlobalCfg.Visibility inside the function itself.

**Review by Ankurk99 — APPROVED** [review]
LGTM! :+1:

**Review by nyrahul — APPROVED** [review]
